tests/infra/logging_app.py

```diff
@@ -336,25 +365,97 @@ def get_receipt(self, node, idx, seqno, view, timeout=3, domain="private"):
+        with primary.client(user if user else self.user) as c:
             table = "private" if priv else "public"
+            url = f"/app/log/{table}?id={log_id}"
+            if self.scope is not None:
+                url += "&scope=" + self.scope
             return c.get(
-                f"/app/log/{table}?id={log_id}", headers=self._get_headers_base()
+                url,
+                headers=self._get_headers_base(),
+            )
+
+    def post_raw_text(self, log_id, msg, log_capture=None, user=None):
+        primary, _ = self.network.find_primary(log_capture=log_capture)
+        with primary.client(user if user else self.user) as c:
+            url = f"/app/log/private/raw_text/{log_id}"
+            if self.scope is not None:
+                url += "?scope=" + self.scope
+            return c.post(
+                url,
+                msg,
+                headers={"content-type": "text/plain"},
             )
+
+
+def scoped_txs(identity, verify=True):
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            network = None
+            node = None
+            previous_txs = None
+            headers = {}
+
+            if isinstance(args[0], infra.network.Network):
+                network = args[0]
+                if hasattr(network, "txs"):
+                    previous_txs = network.txs
+                if hasattr(network, "jwt_issuer") and network.jwt_issuer is not None:
+                    headers = infra.jwt_issuer.make_bearer_header(
+                        network.jwt_issuer.issue_jwt()
+                    )
+                node = network.find_random_node()
+            elif isinstance(args[0], infra.node.Node):
+                node = args[0]
+            else:
+                raise ValueError("invalid args[0]")
+
+            scope = get_fresh_scope(node, identity, headers)
+
+            if network:
+                network.txs = LoggingTxs(identity, scope=scope)
+                network.txs.network = network
+                r = func(*args, **kwargs)
+            else:
+                r = func(*args, **dict(kwargs, scope=scope))
+
+            if network:
+                if verify:
+                    network.txs.verify(network=network)
+                network.txs = previous_txs
+            return r
+
+        def get_fresh_scope(node, identity, headers, attempts=5):
+            while attempts > 0:
+                with node.client(identity) as c:
+                    r = c.get(
+                        f"/app/log/fresh_scope?scope={func.__name__}", headers=headers
+                    )
+                    if r.status_code == http.HTTPStatus.OK:
+                        return r.body.json()["scope"]
+                    elif r.status_code == http.HTTPStatus.NOT_FOUND:
+                        attempts -= 1
+                        time.sleep(0.1)
+
+            raise ValueError("fresh scope request failed")
```

Comment:
Shouldn't `func.__name__` give each call a suitably unique scope name already? Why do we have to additionally call `fresh_scope`?

Reply:
If we can avoid calling `GET /fresh_scope`, and reconstructing `LoggingTxs`, then we don't need to pass `identity`, and the decorators become `@scoped_txs()` rather than `@scoped_txs("user0")`. I think that's more readable (I assumed `"user0"` was the scope prefix!), and I'd lean towards giving `identity` a default argument of `"user0"` rather than passing it repeatedly.

Reply:
The function name is not enough, e.g. the full test suite runs some tests multiple times and we want separate scopes for each of them. Post-hoc I found that we could possible do that from Python by checking that the size of the table is `0` instead of adding the `fresh_scope` endpoint. I don't have a preference for one over the other though.

For the user, I don't know whether `user0` is mostly/always the expected argument, but it was pretty much the choice for all the tests I remember changing.

Reply:
+1 for removing fresh_scope and defaulting user0.